06-building-a-program-start-to-finish/solving-the-problem/output/InvertedIndex.js

```diff
@@ -0,0 +1,133 @@
+*/
+  constructor() {
+    this.mainIndex = {};
+    this.bookNames = [];
+  }
+
+/**
+ * Checks input to see if it comforms to specific standards.
+ * @param {Array} book - Book file to be validated
+ * @returns {boolean} true/false - returns validation status.
+ * @memberOf InvertedIndex
+ */
+  validateInput(book) {
+    if (Array.isArray(book) && book.length > 0 && typeof book[0] === 'object') {
+      if (book[0].text && book[0].title) {
+        if (typeof (book[0].text) === 'number') {
+          return false;
+        }
+        return true;
+      }
+    }
+    return false;
+  }
+
+
+/**
+ * Gets a book and returns it as as a concatenated text
+ * @param {Object} book - a book object with title and text property
+ * @returns {String} text- returns boolean if conditions are not met
+ * @memberOf InvertedIndex
+ */
+  getBookText(book) {
+    const status = this.validateInput([book]);
+    if (status) {
+      this.bookNames.push(book.title);
+      return `${book.title} ${book.text}`;
+    }
+    return status;
+  }
+
+/**
+* Removes characters, whitespaces and converts text to array elements.
+* @param {String} text returned from getBookAsText
+* @returns {Array} -returns an array of words in lower-cases with no characters
+* @memberOf InvertedIndex
+*/
+  generateToken(text) {
+    return text.toLowerCase()
+    .replace(/[^\w\s]|_/g, '')
+    .split(/\s+/);
+  }
+
+
+/**
+ * Gets an array of Words and makes element have unique occurences
+ * @param {Object} Words - a book object with title and text property
+ * @returns {Array} Words - a filtered array with unique elements
+ * @memberOf InvertedIndex
+ */
+  createUniqueWords(Words) {
+    return Words.filter((element, index) =>
+        Words.indexOf(element) === index);
+  }
+
+
+/**
+ * Builds an index for a Book Objects
+ * @param {Array} books - An Array of book objects
+ * @returns {String} Build status as a feedback message
+ * @memberOf InvertedIndex
+ */
+  buildIndex(books) {
+    let nonUniqueWords = '';
+    let words = '';
+    books.forEach((book) => {
+      nonUniqueWords = this.generateToken(this.getBookText(book));
+      words = this.createUniqueWords(nonUniqueWords);
+      words.forEach((word) => {
+        this.addWordToMainIndex(word, book.title);
+      });
+    });
+    return 'Index Built';
+  }
+
+
+/**
+ * Takes a word and a book title and stores it in tne mainindex
+ * @param {String} word
+ * @param {String} bookTitle
+ * @returns {none} ...
+ * @memberOf InvertedIndex
+ */
+  addWordToMainIndex(word, bookTitle) {
+    if (this.mainIndex[word]) {
+      this.mainIndex[word].push(bookTitle);
+    } else {
+      this.mainIndex[word] = [bookTitle];
+    }
+  }
+
+/**
+ *
+ * Takes in word(s) and returns found results based on created index
+ * @param {String} searchedWords - Word(s) used to initiate a search
+ * @returns {Array} searchResult - An array of matched books
+ * @memberOf InvertedIndex
+*/
+  searchIndex(searchedWords) {
```

Comment:
I still see no implementation of search for multiple files Rowland.

I expect that I'll be able to pass in the file name to search through a particular file and then maybe 'all' to search through all files.